Add capabilities

baylesj opened the pull request:
This patch adds specifications for getting device capabilities through the open screen protocol. Specifically, what types of actions and media does the receiver support?

NOTE: This patch is based on @pthatcherg's remote playback pull request, so make sure to exclude those commits when diffing!

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+  request
+  1: [* text] ; urls
+  2: microseconds ; watch-duration
+  3: int ; watch-id
```
Seems be better to use `uint`instead of `int` for consistency with `connection-id`, `request-id`

btolsch replied:
This comment and your others apply to Peter's PR (#134).  Unfortunately, it seems like GitHub doesn't have support for dependent PRs (or maybe just not from separate authors?).  Looking at the individual commits is easier for me than looking at the total diff.

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+; type key 18
+remote-playback-availability-response = {
+  response
+  url-availabilities: [* url-availability]
```
integer key to minimize payload

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+    current: 2
+    future: 3
+    enough: 4
+  ) ; loaded
```
Even if `3` and `4`'s types aren't used elsewhere, I think pulling them out into enums defined outside this map would be more readable. Plus the generated C++ code would be more readable and easier to use

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+    descriptions
+    chapters
+    metadata
+  ) ; kind
```
same as above - define the enum as a type outside the map

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+track-state = (
+  1: text ; id
+  2: text ; label
+  3: text ; language
```
Is language shared with `video-track-state` too?

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+
+video-track-state = {
+  track-state
+  5 : bool ; selected
```
looks like this should be `4` not `5`

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+
+track-track-state = {
+  track-state
+  6 : mode ; text-track-mode
```
looks like this should be `4` not `6`

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+}
+
+; type key 119
+remote-playback-termination-event = {
```
Do we still need ` 0: request-id ; request-id` added here, even if isn't a request or response?

rwkeane commented on index.bs:
```diff
@@ -843,4 +1380,227 @@ result = (
+    captions
+    descriptions
+    chapters
+    metadata
```
Give integer values

btolsch commented on index.bs:
```diff
@@ -284,31 +284,55 @@ mechanism for metadata discovery.
-the IP and port from the SRV record.  Prior to authentication, a message may be
-exchanged (such as further metadata), but such info should be treated as
-unverified (such as indicating to a user that a display name of an
+If a listening agent ("controller") wants to connect to an advertising agent
```
I'm sure Peter and/or Mark will chime in, but I think they have avoided controller/receiver terminology is because it seems Presentation-API-centric.  I'm neutral about this though.

baylesj replied:
I'll leave as is for now, and bring it up with them.

pthatcherg replied:
I like controller/receiver as terms in general, but in the context of mDNS and discovery, we can't assume the receiver is always the advertiser and the controller is always the listener.

baylesj replied:
%s/controller/listener/g
%s/receiver/advertiser/g

btolsch commented on index.bs:
```diff
@@ -284,31 +284,55 @@ mechanism for metadata discovery.
 
+: audio-only (optional)
+:: Generally speaking, receivers are assumed to support streaming both
+audio and video. If an receiver wishes to indicate that they only support audio
```
s/an receiver/a receiver/

aside: Oh man, did GitHub remove the suggestion feature?  That was really nice for typos like this.

btolsch commented on index.bs:
```diff
@@ -284,31 +284,55 @@ mechanism for metadata discovery.
+<!-- TODO(jophba): figure out the units for this -->
+: max-pixel-fill-rate (optional)
+:: In combination with the supported resolutions, the receiver has the
+option to to reply with the supported pixel fill rate in <UNITS>.
```
s/to to/to/

baylesj replied:
Thx

btolsch commented on index.bs:
```diff
@@ -1206,11 +1198,18 @@ resolution = {
-<!-- TODO(jophba): flag enum? -->
 codec = (
-
+  H-264-1080
```
These need to be `name: number`.

btolsch commented on index.bs:
```diff
@@ -1206,11 +1198,18 @@ resolution = {
 codec = (
-
+  H-264-1080
+  H-264-720
```
Why is there a separate 720 and 1080 h264?

pthatcherg commented on index.bs:
```diff
@@ -141,7 +141,7 @@ Presentation API Requirements {#requirements-presentation-api}
     {{PresentationConnection}} to an existing presentation on the
     receiver, given its [=presentation request URL=] and [=presentation ID=].
 
-6.  It must be possible to to close a {{PresentationConnection}} between a
```
About all these typos you found and fixed: could you make a separate PR that fixes those?  It would be nice to have them fixed independently of this PR.

baylesj replied:
https://github.com/webscreens/openscreenprotocol/pull/151

pthatcherg commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
-To learn further metadata, an agent may send an agent-info-request
-message (see [[#appendix-a]]) and receive back an agent-info-response message.  The
-messages may contain the following information with the following meaning:
+To learn further metadata, the controller may send an agent-info-request
```
Actually, both sides will probably send this message to learn about the other side.

baylesj replied:
I'll cleanup language.

pthatcherg commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
 
 : model-name (optional)
-:: If the agent is a hardware device, the model name of
+:: If the receiver is a hardware device, the model name of
```
The device might be a controller.  Imagine a TV wants to connect back to controller devices.

pthatcherg commented on index.html:
```diff
@@ -1696,30 +1698,85 @@ <h2 class="heading settled" data-level="3" id="discovery"><span class="secno">3.
+   <p>To learn further metadata, the controller may send an agent-info-request
+message (see <a href="#appendix-a">Appendix A: Messages</a>) and receive back an agent-info-response message.
+The messages may contain the following properties:</p>
    <dl>
```
Same here: we can't assume the agent is a controller.

pthatcherg commented on index.html:
```diff
@@ -1540,7 +1540,9 @@ <h3 class="heading settled" data-level="1.1" id="terminology"><span class="secno
+a browser). <a data-link-type="dfn" href="https://w3c.github.io/presentation-api/#dfn-presentation-display-availability" id="ref-for-dfn-presentation-display-availability">presentation display availability</a> refers to whether or not a <a data-link-type="dfn" href="https://w3c.github.io/presentation-api/#dfn-receiver" id="ref-for-dfn-receiver①">receiver</a> is compatible with a <a data-link-type ...
+browsing context within the browser, whereas here the "controller" refers to the
+browser itself, although it may be acting on behalf of a browsing context.</p>
    <p>We borrow terminology from the <a data-link-type="biblio" href="#biblio-remote-playback">Remote Playback API</a>.  The
```
Hmm... do you need to rebase again?  I'm seeing in your diff things I made in the last patch of my PR.

baylesj replied:
I think the PR's diff is just bunk, the log looks right to me.

pthatcherg commented on index.html:
```diff
@@ -1696,30 +1698,85 @@ <h2 class="heading settled" data-level="3" id="discovery"><span class="secno">3.
     <dt data-md>model-name (optional)
     <dd data-md>
-     <p>If the agent is a hardware device, the model name of
+     <p>If the receiver is a hardware device, the model name of
```
Here again with assuming receiver == hardware

pthatcherg commented on index.html:
```diff
@@ -1696,30 +1698,85 @@ <h2 class="heading settled" data-level="3" id="discovery"><span class="secno">3.
-make UI affordance for indicating to the user that the display name has
-changed.</p>
+     <p>The display name of the receiver, intended to be displayed to a user by the
+ controller. The controller should indicate through the UI if the receiver
```
Either side can get the display name of other; it's mutual.  It's not just receiver -> controller.

pthatcherg commented on index.bs:
```diff
@@ -1527,10 +1579,64 @@ agent-info-response = {
+bit-rate-limits = {
+ 0: uint ; min-bit-rate
+ 1: uint ; max-bit-rate
+ 2: uint ; max-latency
```
I don't think these should be bundled together.  It's much more common to want a max than a min bitrate.

pthatcherg commented on index.bs:
```diff
@@ -1527,10 +1579,64 @@ agent-info-response = {
+bit-rate-limits = {
+ 0: uint ; min-bit-rate
+ 1: uint ; max-bit-rate
+ 2: uint ; max-latency
```
max-latency?  I don't know what that means and it seems very hard to define.  Perhaps we can address something like the sender being able to control the receiver's jitter buffer, but it doesn't seem like a capability at all.

baylesj replied:
Deleting.

baylesj commented on index.html:
```diff
@@ -2351,10 +2790,64 @@ <h2 class="heading settled" id="appendix-a"><span class="content">Appendix A: Me
+ 0: uint ; min-bit-rate
+ 1: uint ; max-bit-rate
+ 2: uint ; max-latency
+}
```
I'll split up.

baylesj commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+:: The receiver may indicate the practical audio bit rate limits, including the
+    minimum and maximum bit rates, and the maximum latency. Physical hardware
+    requirements, such as maximum decoder throughput or maximum audio buffer
+    size inform these values.
```
Removing audio bit rate.

baylesj commented on index.bs:
```diff
@@ -1527,10 +1579,64 @@ agent-info-response = {
+
+streaming-capabilities = {
+  0: uint ; max-pixel-fill-rate
+  1: [1* codec] ; supported-codecs
```
Changing name.

baylesj commented on index.html:
```diff
@@ -2351,10 +2790,64 @@ <h2 class="heading settled" id="appendix-a"><span class="content">Appendix A: Me
+
+streaming-capabilities = {
+  0: uint ; max-pixel-fill-rate
+  1: [1* codec] ; supported-codecs
```
String sounds great.

baylesj commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+: max-pixel-fill-rate (required)
+:: The receiver shall reply with the supported pixel fill rate in megapixels per
+    second. Requesting agents can use this information to make better-educated
+    decisions about what resolution and frame rates to stream content at.
```
Added new codec type.

baylesj commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+If the receiver is already authenticated, the controller has the ability to
+request additional information by sending an streaming-capabilities-request
+message, and receive back a streaming-capabilities-response message with the
+following properties:
```
I'll push a new revision addressing your feedback, then we can pull over.

mfoltzgoogle replied:
codecs could be used for streaming or remote playback, so I think we should keep it generic.  Unless we believe that the capabilities will be different for a given device.

baylesj replied:
Removed streaming capabilities, will fixup other patch.

baylesj commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+:: Generally speaking, receivers are assumed to support streaming both
+    audio and video. If a receiver wishes to indicate that they only support audio
+    stream, they can pass this property as true. In that case, the controller
+    should indicate through the UI that the receiver is audio only.
```
Agreed.

baylesj commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+:: The receiver may indicate the practical audio bit rate limits, including the
+    minimum and maximum bit rates, and the maximum latency. Physical hardware
+    requirements, such as maximum decoder throughput or maximum video buffer
+    size inform these values.
```
Removed audio bit rate limits.

mfoltzgoogle commented on index.bs:
```diff
@@ -297,32 +297,84 @@ mechanism for metadata discovery.
+treated as unverified (such as indicating to a user that a display name of an
+unauthenticated agent is unverified). Note that both listeners and advertisers
+may request information about each other (performing as a "requester" to
+query a "requestee").
```
This paragraph is vague and confusing.  It starts by suggesting that the listener wants to learn about the advertiser.  But then says that each side can learn about the other once they are connected.   It's also unclear what message in the spec can be sent as "unverified" data.

I would suggest dropping the last sentence as it's covered in a more specific way in the next paragraph.

baylesj replied:
Agreed.

mfoltzgoogle commented on index.bs:
```diff
@@ -297,32 +297,84 @@ mechanism for metadata discovery.
-messages may contain the following information with the following meaning:
+To learn further metadata, devices may send an agent-info-request
+message (see [[#appendix-a]]) and receive back an agent-info-response message.
+Note that both the listener and advertiser may send this request to learn about
```
s/listener and advertiser/any agent/

baylesj replied:
Changed.

mfoltzgoogle commented on index.bs:
```diff
@@ -297,32 +297,84 @@ mechanism for metadata discovery.
+Note that both the listener and advertiser may send this request to learn about
+the capabilities of the other device.
+
+The messages may contain the following properties:
```
s/may// since there are required properties

mfoltzgoogle replied:
Are these properties of the agent-info-request or the agent-info-response?

baylesj replied:
Clarified.

mfoltzgoogle commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+:: The receiver must indicate to the controller what codecs it understands for
+    remote playback.
+
+: min-resolution (optional)
```
+1  I'm not sure codecs belong in a device-level metadata, but instead should be part the negotiation specifically for streaming or remote playback.  Ideally we can share some messages between the two because there's overlap in the use case: the controller has some media it wants to play on the receiver with a given codec, resolution/framerate, and bitrate.  The receiver can answer if it can play that media at all and/or play it smoothly or efficiently (see the [Media Capabilities API](https://wicg.github.io/media-capabilities/) for a Web API for querying this data).

mfoltzgoogle commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+: aspect-ratio (optional)
+:: The receiver may indicate what its ideal aspect ratio is, e.g. a 16:10
+    display could return this value as 1.6 to indicate its preferred content
+    scaling.
```
Nit, but this feels better as a rational number, because of 16:9, 4:3, etc.

mfoltzgoogle commented on index.bs:
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
+    sRGB v4. The controller may use these values to determine how to encode
+    video.
+
+: supports-scaling (optional)
```
In general there are multiple native resolutions that have hardware support for scaling.  The receiver could list a set of supported resolutions and then indicate if non-standard resolutions are also supported.

mfoltzgoogle commented on index.bs:
```diff
@@ -297,32 +297,40 @@ mechanism for metadata discovery.
+: supports-video (optional)
+:: The requestee has to indicate that it supports video streaming. If false
+    or not included, it is assumed video streaming is not supported.
+
```
In the future we should consider capabilities for audio input and video input as well.  But these capabilities are more often associated with removable peripherals, so it may not be suitable for device-level metadata.

Maybe we should make all capabilities playback-level metadata?  That is, part of the url-availability-request, media-request and/or remote-playback-availability-request?

baylesj replied:
That's a good question. I'll leave this for now, I'll be in Fremont tomorrow/Friday if you want to sync up in person and hammer down some of the specifics around this.

rektide replied:
This is an even bigger can of worms, but I'd like to see input device support as a capability too. The trendy new big touchscreen collaborative displays come to mind as the sort of thing one would expect to openscreenprotocol into, which would have their own input device systems.

I don't know whether this capability ought be broken down into smaller pieces- `supports-input` or `supports-touch` `supports-keyboard` `supports-mouse`?

As per previous comments, maybe doesn't need to be done now? I can open a follow-up item depending on peoples feels here.

baylesj replied:
I would vote for a follow up item, personally. I generally prefer smaller patches when possible.

pthatcherg commented on index.bs:
```diff
@@ -297,32 +297,37 @@ mechanism for metadata discovery.
+Any agent may send this request to learn about the capabilities of
+another device.
+
+The agent-info-response messages contain the following properties:
```
"The X message contains" or "X messages contain", not "The X messages contain,"

baylesj replied:
+1

pthatcherg commented on index.bs:
```diff
@@ -297,32 +297,37 @@ mechanism for metadata discovery.
-    make UI affordance for indicating to the user that the display name has
-    changed.
+:: The display name of the agent, intended to be displayed to a user by the
+     requester. The requester should indicate through the UI if the device
```
"if the device" => "if the responder"

baylesj replied:
+1

pthatcherg commented on index.bs:
```diff
@@ -297,32 +297,37 @@ mechanism for metadata discovery.
 
-<!-- TODO: Add device type and/or capabilities -->
+: supports-audio (optional)
+:: The agent has to indicate that it supports audio streaming. If false
```
It think we should show support for audio and video in general, not specific to streaming (such as for the presentation API).  This is mostly to know if we should show an a/v icon or just an audio icon.  

Also, the name should probably be "receives-audio" and "receives-video".

baylesj replied:
Agreed.

baylesj commented:
@mfoltzgoogle, thoughts on closed comments below. TL;DR I believe I have addressed all of your comments in my other patch,
https://github.com/webscreens/openscreenprotocol/pull/154

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272304956**
"Codecs should be generic":
_Agreed, cleaned up in second patch._

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272308357**
"Codecs belong in streaming":
_Agreed, moved as part of moving to second patch._

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272308928**
"Rational number":
_Makes sense. Added a new ratio type, since we don't appear to have one._

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272309518**
"Multiple native resolutions":
_Agreed, I updated to add a : native-resolutions (optional) field._

mfoltzgoogle commented:
Approved the current PR with two items for follow-up discussion:

- In the `agent-info` message, using booleans versus a set-of-enums to describe capabilities
- How to describe the semantics of `receives-audio` and `receives-video` more specifically.  Does that mean an agent is required to process set of messages, a set of codecs or is it entirely up to the implementation?

Still awaiting approval from @pthatcherg.

pthatcherg commented:
I think enums instead of bools is an interesting idea.  

As for "receives-audio" and "receives-video", I think the part we're missing is to have a capabilities for indicating what parts of the protocol the agent speaks.  For example, it acts as a receiver for presentation API or remote playback API or streaming or all of them.  Or it acts as a sender/controller for presentation API or remote playback API or streaming or all of them.

I think "receives-audio" and "receives-video" implies that it's a receiver, and for now that probably means both presentation API and remote playback API, but not streaming.

pthatcherg commented:
I think we can land this and continue discussing in more detail.
